### PR TITLE
OTLP exporter: Fix TypeInitializationException thrown by exporter

### LINKED ISSUE #1854: Google.Protobuf 3.15.0 breaks the Otlp Exporter
# Bug Report

A bug report is not really the right way to report this one, this is more a heads up I guess if this ever manifests when updating the dependencies.

List of [NuGet packages](https://www.nuget.org/profiles/OpenTelemetry) and
version that you are using (e.g. `OpenTelemetry 0.4.0-beta.2`): 

* `<PackageReference Include="OpenTelemetry" Version="1.0.1" />`
* `<PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.0.1" />`
* `<PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.0.1" />`

Runtime version:

* `net5.0`

## Symptom

Upgrading Google.Protobuf 3.14.0 -> 3.15.0 breaks the Otlp (Batch?) Exporter.

**What is the expected behavior?**

No exception. (So it finds the ByteString constructor)

**What is the actual behavior?**

The application crashes with this exception.

```
Unhandled exception. System.TypeInitializationException: The type initializer for 'OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ActivityExtensions' threw an exception.
 ---> System.ArgumentNullException: Value cannot be null. (Parameter 'con')
 at System.Reflection.Emit.DynamicILGenerator.Emit(OpCode opcode, ConstructorInfo con)
 at OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ActivityExtensions.CreateByteStringCtorFunc()
 at OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ActivityExtensions..cctor()
 --- End of inner exception stack trace ---
 at OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ActivityExtensions.AddBatch(ExportTraceServiceRequest request, Resource processResource, Batch`1& activityBatch)
 at OpenTelemetry.Exporter.OtlpTraceExporter.Export(Batch`1& activityBatch)
 at OpenTelemetry.BatchExportProcessor`1.ExporterProc()
 at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
 at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
 at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
 at System.Threading.ThreadHelper.ThreadStart()
```

## Reproduce

Update your application to Google.Protobuf 3.15.0 (local to the project)

## Additional Context

The new Google.Protobuf 3.15.0 package changed the ByteString type thus breaking compat with this package it seems.
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs#L346 (here)

https://github.com/protocolbuffers/protobuf/pull/7645

### PR DESCRIPTION
**Pull request by alanwest**
Fixes #1854.

Uses new `UnsafeByteOperations.UnsafeWrap` method in Google.Protobuf 3.15+ to avoid additional `byte[]` allocation.

Perf before:

|                Method | NumberOfBatches | NumberOfSpans |        Mean |     Error |    StdDev |       Gen 0 |       Gen 1 |      Gen 2 |  Allocated |
|---------------------- |---------------- |-------------- |------------:|----------:|----------:|------------:|------------:|-----------:|-----------:|
| OtlpExporter_Batching |               1 |         10000 |    83.97 ms |  1.561 ms |  1.533 ms |   4166.6667 |   1500.0000 |   500.0000 |   21.29 MB |
| OtlpExporter_Batching |              10 |         10000 |   878.07 ms | 16.511 ms | 27.129 ms |  47000.0000 |  19000.0000 |  7000.0000 |  212.88 MB |
| OtlpExporter_Batching |             100 |         10000 | 9,006.40 ms | 49.816 ms | 44.160 ms | 509000.0000 | 201000.0000 | 89000.0000 | 2128.82 MB |

Perf after:

|                Method | NumberOfBatches | NumberOfSpans |        Mean |      Error |     StdDev |       Gen 0 |       Gen 1 |      Gen 2 |  Allocated |
|---------------------- |---------------- |-------------- |------------:|-----------:|-----------:|------------:|------------:|-----------:|-----------:|
| OtlpExporter_Batching |               1 |         10000 |    83.43 ms |   1.487 ms |   1.318 ms |   4500.0000 |   1833.3333 |   666.6667 |   21.67 MB |
| OtlpExporter_Batching |              10 |         10000 |   872.01 ms |  13.699 ms |  11.440 ms |  45000.0000 |  18000.0000 |  7000.0000 |   216.7 MB |
| OtlpExporter_Batching |             100 |         10000 | 9,164.78 ms | 122.865 ms | 108.917 ms | 547000.0000 | 222000.0000 | 99000.0000 | 2166.95 MB |

Seems there is a bit of an increase in memory allocated... I will investigate why.

### CODE REVIEW

**Comment by alanwest** [comment]
@CodeBlanch @cijothomas @reyang

I fairly certain the increase in memory allocated is not due to the code changes here. The new `UnsafeByteOperations.UnsafeWrap` method avoids an additional allocation of a byte array. The original code was achieving this by reflection which breaks consumers of the Google.Protobuf package version 3.15+.

I suspect that the increase in allocation is related to upgrading our dependency to 3.15.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,12 @@ please check the latest changes
 
 ## Unreleased
 
+* Resolves `System.TypeInitializationException` exception when using the
+  exporter with an application that references Google.Protobuf 3.15. The OTLP
+  exporter now depends on Google.Protobuf 3.15.5 enabling the use of the new
+  `UnsafeByteOperations.UnsafeWrap` to avoid unnecessary allocations.
+  ([#TBD](https://github.com/open-telemetry/opentelemetry-dotnet/pull/TBD))
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md:16]:
```suggestion
  ([#1873](https://github.com/open-telemetry/opentelemetry-dotnet/pull/1873))
```

**Review by reyang — APPROVED** [review]
LGTM.

**Comment by CodeBlanch** [comment]
LGTM I don't think the change is the reason for the perf hit:

https://github.com/protocolbuffers/protobuf/blob/e9360dfa53f151e64c996823785e0eb6c8118620/csharp/src/Google.Protobuf/UnsafeByteOperations.cs#L78

https://github.com/protocolbuffers/protobuf/blob/e9360dfa53f151e64c996823785e0eb6c8118620/csharp/src/Google.Protobuf/ByteString.cs#L65

That all seems perfectly reasonable to me. Assuming it is flowing through that ReadOnlyMemory<byte> due to implicit cast from byte[].

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md (hunk 1 above)
Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md]: Good catch, thanks!

**Pull request merged**